packages/react/src/Header/Header.tsx

```diff
@@ -4,68 +4,131 @@ import {get} from '../constants'
-  align-items: center;
-  flex-wrap: nowrap;
-  overflow: auto;
-
-  ${sx};
-`
-const HeaderItem = styled.div<StyledHeaderItemProps>`
-  display: flex;
-  margin-right: ${get('space.3')};
-  align-self: stretch;
-  align-items: center;
-  flex-wrap: nowrap;
-
-  ${({full}) =>
-    full &&
-    css`
-      flex: auto;
-    `};
-
-  ${sx};
-`
+type StyledHeaderProps = React.ComponentProps<'header'> & SxProp
+type StyledHeaderItemProps = React.ComponentProps<'div'> & SxProp & {full?: boolean}
+type StyledHeaderLinkProps = React.ComponentProps<'a'> & SxProp & {to?: Location | Pathname}
 
-HeaderItem.displayName = 'Header.Item'
+const CSS_MODULES_FEATURE_FLAG = 'primer_react_css_modules_team'
 
-const HeaderLink = styled.a.attrs<StyledHeaderLinkProps>(({to}) => {
-  const isReactRouter = typeof to === 'string'
-  if (isReactRouter) {
-    // according to their docs, NavLink supports aria-current:
-    // https://reacttraining.com/react-router/web/api/NavLink/aria-current-string
-    return {'aria-current': 'page'}
-  } else {
-    return {}
-  }
-})<StyledHeaderLinkProps>`
-  font-weight: ${get('fontWeights.bold')};
-  color: ${get('colors.header.logo')};
-  white-space: nowrap;
-  cursor: pointer;
-  text-decoration: none;
-  display: flex;
-  align-items: center;
-
-  &:hover,
-  &:focus {
+const StyledHeader = toggleStyledComponent(
+  CSS_MODULES_FEATURE_FLAG,
+  'header',
+  styled.header<StyledHeaderProps>`
+    z-index: 32;
+    display: flex;
+    padding: ${get('space.3')};
+    font-size: ${get('fontSizes.1')};
+    line-height: ${get('lineHeights.default')};
     color: ${get('colors.header.text')};
-  }
+    background-color: ${get('colors.header.bg')};
+    align-items: center;
+    flex-wrap: nowrap;
+    overflow: auto;
+
+    ${sx};
+  `,
+)
+
+const Header = React.forwardRef<HTMLElement, StyledHeaderProps>(function Header(
+  {children, className, ...rest},
+  forwardRef,
+) {
+  const enabled = useFeatureFlag(CSS_MODULES_FEATURE_FLAG)
+  return (
+    <StyledHeader ref={forwardRef} className={clsx(className, {[classes.Header]: enabled})} {...rest}>
+      {children}
+    </StyledHeader>
+  )
+}) as PolymorphicForwardRefComponent<'header', StyledHeaderProps>
+
+Header.displayName = 'Header'
+
+const StyledHeaderItem = toggleStyledComponent(
+  CSS_MODULES_FEATURE_FLAG,
+  'div',
+  styled.div<StyledHeaderItemProps>`
+    display: flex;
+    margin-right: ${get('space.3')};
+    align-self: stretch;
+    align-items: center;
+    flex-wrap: nowrap;
+
+    ${({full}) =>
+      full &&
+      css`
+        flex: auto;
+      `};
+
+    ${sx};
+  `,
+)
+
+const HeaderItem = React.forwardRef<HTMLElement, StyledHeaderItemProps>(function HeaderItem(
+  {children, className, ...rest},
+  forwardRef,
+) {
+  const enabled = useFeatureFlag(CSS_MODULES_FEATURE_FLAG)
+  return (
+    <StyledHeaderItem
+      ref={forwardRef}
+      className={clsx(className, enabled && classes.HeaderItem, enabled && rest.full && classes.Full)}
+      {...rest}
```

Comment:
For variant type props, we're using data-attributes for styling instead of classnames. You can read a little bit about that in the [CSS authoring guide](https://github.com/primer/react/blob/main/contributor-docs/authoring-css.md#component-prop-variants-as-data-attributes).

So for the `full` prop, instead of a classname it would look something like this:

```suggestion
      className={clsx(className, enabled && classes.HeaderItem}
      data-full={full}
      {...rest}
```

And then in CSS 

```
&:where([data-full]) {
  ....
}

If you check out the ButtonBase component you can see this in action 😄

Reply:
Nice, makes a lot of sense, updated!